Lib/dataclasses.py

```diff
@@ -1309,6 +1309,14 @@ def _asdict_inner(obj, dict_factory):
         # generator (which is not true for namedtuples, handled
         # above).
         return type(obj)(_asdict_inner(v, dict_factory) for v in obj)
+    elif isinstance(obj, dict) and hasattr(obj, 'default_factory'):
```

Comment:
I think this test should be `type(obj)(getattr(obj, 'default_factory'))` to match the `result =` line below. And it could use a walrus operator to only calculate it once.

Reply:
Never mind on the comment about the walrus operator. But I think the hasattr test should be `hasattr(type(obj), 'default_factory')`. I'm open to debate on it, though.

Reply:
Yes, I agree `hasattr(type(obj), 'default_factory')` would bit more robust as it rules out users modifying/adding this attribute.